### PR TITLE
Experiments public contract

### PR DESCRIPTION
**Pull request by wzieba**
### Description

This PR introduces improvements to the public contract:
- extracts an interface: `VariationsRepository`, so the clients can easily double this in tests + KDoc comments
- introduces a behavior change: refreshing cache on SDK initialization, as I believe this is something that every client would do on initialization anyway
- simplifies `refresh` API (no need for two methods)
- adds `README.md` with an explanation about how to use the SDK

### Testing

The only behavior change has been covered  with automated tests. It's still possible to smoke test using instructions from #241 , if necessary.

### CODE REVIEW

File: experimentation/src/main/java/com/automattic/android/experimentation/ExPlat.kt [hunk 1]
```diff
@@ -5,50 +5,38 @@ import com.automattic.android.experimentation.ExPlat.RefreshStrategy.IF_STALE
...
 import kotlinx.coroutines.CoroutineScope
-import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.launch
-import java.io.File
 
-class ExPlat internal constructor(
+public class ExPlat internal constructor(
     private val platform: String,
-    private val experiments: Set<Experiment>,
+    experiments: Set<Experiment>,
```
Comment (wzieba) [inline experimentation/src/main/java/com/automattic/android/experimentation/ExPlat.kt:13]: I wonder: is there a value in providing experiments in the constructor and then checking for its existence in `getVariation`? I don't see good benefits 🤔

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -30,10 +30,18 @@ steps:
       cp gradle.properties-example gradle.properties
       ./gradlew test
 
+  - label: "API verification"
+    key: "api"
+    plugins: [$CI_TOOLKIT]
+    command: |
+      cp gradle.properties-example gradle.properties
+      ./gradlew :experimentation:apiCheck
```
Comment (ParaskP7) [inline .buildkite/pipeline.yml:38]: Praise (❤️): Awesome, love it, thanks for adding the `binary-compatibility-validator` plugin into this repo, let's start using this more and more within this project and other repos as well! 🥇

File: experimentation/README.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
+# Experimentation
```
Comment (ParaskP7) [inline experimentation/README.md:1]: Praise (❤️): Thanks for creating a `README.md` to help with the API usage!

File: experimentation/src/main/java/com/automattic/android/experimentation/ExPlat.kt [hunk 4]
```diff
@@ -5,50 +5,38 @@ import com.automattic.android.experimentation.ExPlat.RefreshStrategy.IF_STALE
...
+) : VariationsRepository {
     private val activeVariations = mutableMapOf<String, Variation>()
     private val experimentIdentifiers: List<String> = experiments.map { it.identifier }
 
     private var anonId: String? = null
 
-    fun configure(anonId: String? = null) {
+    init {
+        coroutineScope.launch {
+            refresh()
```
Comment (ParaskP7) [inline experimentation/src/main/java/com/automattic/android/experimentation/ExPlat.kt]: Question (❓): I think this is a good move, but how about when users would like to also use `configure(...)` to setup ExPlat with an anonymous id, they will then need to trigger `refresh()` manually once more, right, or, how will they better use the API then? 🤔

File: experimentation/src/main/java/com/automattic/android/experimentation/VariationsRepository.kt [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import java.io.File
+
+/**
+ * Repository for accessing [Variation]s for [Experiment]s.
+ *
+ * Should be a Singleton, as it keeps track of the "active" [Variation]s. See [getVariation] for more details.
+ */
+public interface VariationsRepository {
```
Comment (ParaskP7) [inline experimentation/src/main/java/com/automattic/android/experimentation/VariationsRepository.kt:20]: Praise (❤️): Love it! 🥇

**Review by ParaskP7 — COMMENTED** [review]
And with that final review, I think I am done with all the review @wzieba , kudos on this work, it will help our product teams setup and start using [ExPlat](https://experiments.a8c.com/experiments) without much overhead, and mostly, without `FluxC`! 🎉 🥇 🚀 

Let's start resolving the comments one-by-one and merge everything to `trunk` when ready, just let me know how to help you from here! 💯

File: experimentation/src/main/java/com/automattic/android/experimentation/VariationsRepository.kt [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            val assignmentsValidator = AssignmentsValidator(SystemClock())
+            return ExPlat(
+                platform = platform,
+                experiments = experiments,
+                logger = logger,
+                coroutineScope = coroutineScope,
+                isDebug = isDebug,
+                assignmentsValidator = assignmentsValidator,
+                repository = assignmentsRepository,
+            )
```
Comment (wzieba) [inline experimentation/src/main/java/com/automattic/android/experimentation/VariationsRepository.kt]: - [ ] https://github.com/Automattic/Automattic-Tracks-Android/pull/241#discussion_r1736393960
- [ ] https://github.com/Automattic/Automattic-Tracks-Android/pull/242#discussion_r1736610369

File: experimentation/src/main/java/com/automattic/android/experimentation/ExPlat.kt (hunk 4 above)
Comment (wzieba) [reply experimentation/src/main/java/com/automattic/android/experimentation/ExPlat.kt]: True, this was not the best idea. I learned that `anonId` is actually a mandatory information ( p1725007206945009-slack-C7HH3V5AS) and I proposed changes to this initialization methods in #245

Comment (ParaskP7) [reply experimentation/src/main/java/com/automattic/android/experimentation/ExPlat.kt]: Thanks for all this work @wzieba , I followed all the conversation on anonymous id needing to become mandatory (and all), plus did a quick review on #245 , the changes makes sense to me (also, thanks @iangmaia for the review)! 🚀

**Pull request merged**